### PR TITLE
Added support for Multipart OpenEXRs.

### PR DESCRIPTION
**Pull request by ggarra13**
This is my real first contribution to tlRender.  This commit adds multipart EXR support to the OpenEXR reader while still supporting multiview EXRs.

### CODE REVIEW

**Comment by darbyjohnston** [comment]
Thanks, this looks very interesting, I should have time next week to look at it in more detail. What images are you using for testing?

**Comment by ggarra13** [comment]
Here's a single image for you to test:
https://mega.nz/file/vKQSBT5T#nUQBv505Y5-cLeyymzCbYRcej-RIL2silULtE8XfzEg

And here's.the full sequence (240 exr frames):
https://mega.nz/file/zHJHhL4b#3Cz0xGt5qaMSNPS9U49cMNwjwbc3OaOePnsO9ID8hLo

**Comment by darbyjohnston** [comment]
If I understand these changes correctly, each part in the OpenEXR file can represent one or more tlRender layers? So you could have a file like this:
* part 0
    * rgba
* part 1
    * diffuse
    * ambient
* part 2
    * depth

Which would become these tlRender layers:
* rgba
* diffuse
* ambient
* depth

I guess reading the depth layer would be fast since it is in it's own part, but reading the diffuse or ambient layer could be slower since they are interleaved within the same part.

**Comment by ggarra13** [comment]
On 9/2/23 22:55, Darby Johnston wrote:
>
> If I understand these changes correctly, each part in the OpenEXR file 
> can represent one or more tlRender layers? So you could have a file 
> like this:
>
>   * part 0
>       o rgba
>   * part 1
>       o diffuse
>       o ambient
>   * part 2
>       o depth
>
> Which would become these tlRender layers:
>
>   * rgba
>   * diffuse
>   * ambient
>   * depth
>
> I guess reading the depth layer would be fast since it is in it's own 
> part, but reading the diffuse or ambient layer could be slower since 
> they are interleaved within the same part.
>
No.  The way I understand it is that for each part you have up to 5 
channels:

     * part 0
            rgbaz
     * part 1
            diffuse.r diffuse,g diffuse.b
     * part 2
           ambocl.r ambocl.g ambocl.b ambocle.a

Each part has a table associated to it which allows MultiPart reader to 
jump to that section immediately.

AFAIK, you never get, say diffuse and ambient together in the same par.

> Message ID: ***@***.***>
>
-- 
Gonzalo Garramuño
***@***.***

**Comment by darbyjohnston** [comment]
OK, well some description in the PR would be nice so I don't have to guess about the functionality.

Have you tested with non-multipart files that have multiple layers to make sure they still work with these changes?

**Comment by ggarra13** [comment]
I tested with a multiview file (single part, multiple channels file) and it worked okay.

**Comment by darbyjohnston** [comment]
I have a number of exr files with layers, I'll try them with these changes.

**Comment by darbyjohnston** [comment]
Hi and sorry for the delay! Hopefully I will have some time next week to test these changes and review the PR.

File: lib/tlIO/OpenEXRRead.cpp [hunk 1]
```diff
@@ -153,76 +154,87 @@ namespace tl
...
-                        case Imf::PixelType::FLOAT:
-                            info.pixelType = imaging::getFloatType(layer.channels.size(), 32);
-                            break;
-                        case Imf::PixelType::UINT:
-                            info.pixelType = imaging::getIntType(layer.channels.size(), 32);
-                            break;
-                        default: break;
+                            const std::string id = string::Format("tl::io::exr::Read {0}").arg(this);
+                            std::vector<std::string> s;
+                            s.push_back(string::Format(
```
Comment (darbyjohnston) [inline lib/tlIO/OpenEXRRead.cpp:175]: Why was this reformatted from the original code?

File: lib/tlIO/OpenEXRRead.cpp [hunk 2]
```diff
@@ -153,76 +154,87 @@ namespace tl
...
+                        
+                        std::vector<Layer> layers = getLayers(header.channels(), channelGrouping);
+                        size_t offset = _info.video.size();
+                        _info.video.resize( offset + layers.size() );
+                        for (size_t i = 0; i < layers.size(); ++i)
                         {
-                            throw std::runtime_error(string::Format("{0}: Unsupported image type").arg(fileName));
+                            layers[i].partNumber = partNumber;
+                            const auto& layer = layers[i];
+                            _layers.push_back( layer );
```
Comment (darbyjohnston) [inline lib/tlIO/OpenEXRRead.cpp:207]: Could you make a pass over the changes to make the formatting consistent? (e.g., no spaces between parentheses)

File: lib/tlIO/OpenEXRRead.cpp [hunk 3]
```diff
@@ -153,76 +154,87 @@ namespace tl
...
+                            }
+                            logSystem->print(id, string::join(s, '\n'));
                         }
-                        if (imaging::PixelType::None == info.pixelType)
+
+                        // Get the tags.
+                        readTags(header, _info.tags);
+
+                        // Get the layers.
+
```
Comment (darbyjohnston) [inline lib/tlIO/OpenEXRRead.cpp:199]: Not sure why this space was added?

File: lib/tlIO/OpenEXRRead.cpp [hunk 4]
```diff
@@ -153,76 +154,87 @@ namespace tl
...
-                            s.push_back(ss2.str());
-                        }
-                        logSystem->print(id, string::join(s, '\n'));
-                    }
+                        const Imf::Header& header = _f->header(partNumber);
 
-                    // Get the tags.
-                    readTags(_f->header(), _info.tags);
+                        // Get the display and data windows.
+                        _displayWindow = fromImath( header.displayWindow());
```
Comment (darbyjohnston) [inline lib/tlIO/OpenEXRRead.cpp:166]: All of these (display window, data window, etc.) are the same across all parts?

**Review by darbyjohnston — CHANGES_REQUESTED** [review]
Sorry again for the delay! I think this looks OK and it would be good to merge to get some more testing for it. My main question is whether each part has the same display and data window? Also a couple of formatting comments.

File: lib/tlIO/OpenEXRRead.cpp (hunk 4 above)
Comment (ggarra13) [reply lib/tlIO/OpenEXRRead.cpp:166]: Yes.

**Comment by ggarra13** [comment]
I added a fix for reading data and display windows on each frame as those can change in each frame.  Also they can change from left and right eye views.

To test, see the openexr-images at:

https://github.com/AcademySoftwareFoundation/openexr-images.git

particularly, the Beachball demo images.

**Comment by darbyjohnston** [comment]
I think this is pretty close, could you do a bit of code formatting from the comments I left in the review and "resolve" them? Thanks!

**Pull request closed without merging**